File: hearts-scala/src/main/scala/funar/hearts/GameEvent.scala
package funar.hearts

import cats._
import cats.implicits._

import Cards._

enum GameEvent {
	case HandDealt(player: Player, hand: Hand)
	case PlayerTurnChanged(player: Player)
	case LegalCardPlayed(player: Player, card: Card)
	case IllegalCardAttempted(player: Player, card: Card)
	case TrickTaken(player: Player, trick: Trick)
	case GameEnded(won: Player)
}

object GameEvent {
	def eventsWinner(events: Iterable[GameEvent]): Option[Player] =
		events.find({ case GameEnded(won) => true
									case _ => false }) match {
										case Some(GameEnded(won)) => Some(won)
										case _ => None
									}
}

enum GameCommand {
	case DealHands(hands: Map[Player, Hand])
	case PlayCard(player: Player, card: Card)
}

object GameM {
	enum Game[A] {
		case PlayValid(player: Player, card: Card, cont: Boolean => Game[A])
		case RecordEvent(event: GameEvent, cont: Unit => Game[A])
		case GetCommand(cont: GameCommand => Game[A])
		case TurnOverTrick(cont: Option[(Trick, Player)] => Game[A])
		case PlayerAfter(player: Player, cont: Player => Game[A])
		case GameOver(cont: Option[Player] => Game[A])
		case Done(result: A)
	}
	import Game._

	def playValidM(player: Player, card: Card) = PlayValid(player, card, Done(_))
	def recordEventM(event: GameEvent) = RecordEvent(event, Done(_))
	def turnOverTrickM = TurnOverTrick(Done(_))
	def playerAfterM(player: Player) = PlayerAfter(player, Done(_))
	def gameOverM = GameOver(Done(_))
}

import GameM.Game

given gameMonad : Monad[GameM.Game[_]] with {
	override def pure[A](result: A): Game[A] = Game.Done(result)
	override def flatMap[A, B](game: Game[A])(next: A => Game[B]): Game[B] = {
		import Game._
		game match {
			case PlayValid(player, card, cont) =>
				PlayValid(player, card, valid => cont(valid).flatMap(next))
			case RecordEvent(event, cont) =>
				RecordEvent(event, _ => cont(()).flatMap(next))
			case GetCommand(cont) =>
				GetCommand(command => cont(command).flatMap(next))
			case TurnOverTrick(cont) =>
				TurnOverTrick(over => cont(over).flatMap(next))
			case PlayerAfter(player, cont) =>
				PlayerAfter(player, player => cont(player).flatMap(next))
			case GameOver(cont) =>
				GameOver(won => cont(won).flatMap(next))
			case Done(result) => next(result)
		}
	}
	override def tailRecM[A, B](a: A)(f: A => Game[Either[A, B]]): Game[B] = {
		
		import Game._
		def loop(a: A): Game[B] =
			f(a).flatMap {
				case Left(a1) => tailRecM(a1)(f)
				case Right(b) => Done(b)
			}
		loop(a)
	}
}

def tableProcessCommandM(command: GameCommand): Game[Option[Player]] = {
	import GameCommand._
	import GameEvent._
	import GameM._
	import gameMonad._
	command match {
		case DealHands(hands) =>
			val records =
				hands.map { case (player, hand) => recordEventM(HandDealt(player, hand)) }.toSeq
			Foldable[Seq].sequence_(records).map { _ => None }
		case PlayCard(player, card) =>
			playValidM(player, card) >>= { valid =>
				if (valid) {
					recordEventM(LegalCardPlayed(player, card)) >>
					turnOverTrickM >>= { turnOverTrick =>
						turnOverTrick match {
							case Some((trick, trickTaker)) =>
								recordEventM(TrickTaken(trickTaker, trick)) >>
								gameOverM >>= { gameOver =>
									gameOver match {
										case Some(winner) =>
											for {
												_ <- recordEventM(GameEnded(winner))
											} yield Some(winner)
										case None =>
											recordEventM(PlayerTurnChanged(trickTaker)) >> pure(None)
									}
								}
							case None =>
								playerAfterM(player) >>= { nextPlayer =>
									recordEventM(PlayerTurnChanged(nextPlayer)).map { _ => None }
								}
						}
					}
				} else 
					recordEventM(IllegalCardAttempted(player, card)) >> pure(None)
			}
		}
	}

def tableLoopM(command: GameCommand): Game[Option[Player]] =
	tableProcessCommandM(command) >>= {
		case None => Game.GetCommand(tableLoopM)
		case Some(winner) =>
			Game.Done(Some(winner))
	}

File: hearts/src/GameEvent.hs
module GameEvent where

import Cards

import qualified Data.Map as Map
import Data.Map (Map)

data GameEvent
  = HandDealt Player Hand
  | PlayerTurnChanged Player
  | LegalCardPlayed Player Card
  | TrickTaken Player Trick
  | IllegalCardAttempted Player Card
  | GameEnded Player
  deriving (Show, Eq)

data GameCommand
  = DealHands (Map Player Hand)
  | PlayCard Player Card
  deriving (Show, Eq)

-- | Steht in den Events, wer gewonnen hat?
-- >>> let mike = Player "Mike"
-- >>> let peter = Player "Peter"
-- >>> eventsWinner []
-- Nothing
-- >>> eventsWinner [PlayerTurnChanged mike]
-- Nothing
-- >>> eventsWinner [PlayerTurnChanged peter, GameEnded mike]
-- Just (Player {playerName = "Mike"})
eventsWinner :: [GameEvent] -> Maybe Player
eventsWinner [] = Nothing
eventsWinner (first : rest) =
  case first of
    GameEnded winner -> Just winner
    _ -> eventsWinner rest

data Game a =
   PlayValid Player Card (Bool -> Game a)
 | RecordEvent GameEvent (() -> Game a)
 | GetCommand (GameCommand -> Game a)
 | TurnOverTrick (Maybe (Trick, Player) -> Game a)
 | PlayerAfter Player (Player -> Game a)
 | GameOver (Maybe Player -> Game a)
 | Done a

playValidM :: Player -> Card -> Game Bool
playValidM player card = PlayValid player card Done

recordEventM :: GameEvent -> Game ()
recordEventM event = RecordEvent event Done

turnOverTrickM :: Game (Maybe (Trick, Player))
turnOverTrickM = TurnOverTrick Done

playerAfterM player = PlayerAfter player Done

gameOverM = GameOver Done

instance Functor Game

instance Applicative Game

instance Monad Game where
  return = Done
  (PlayValid player card cont) >>= next =
    PlayValid
      player
      card
      ( \valid ->
          cont valid >>= next
      )
  (RecordEvent event cont) >>= next =
    RecordEvent
      event
      ( \() ->
          cont () >>= next
      )
  (GetCommand cont) >>= next =
    GetCommand
      ( \command ->
          cont command >>= next
      )
  (TurnOverTrick cont) >>= next =
    TurnOverTrick
      ( \over ->
          cont over >>= next
      )
  (PlayerAfter player cont) >>= next =
    PlayerAfter
      player
      ( \player ->
          cont player >>= next
      )
  (GameOver cont) >>= next =
    GameOver
      ( \won ->
          cont won >>= next
      )
  (Done result) >>= next = next result

-- | Einen Spielschritt verarbeiten
tableProcessCommandM :: GameCommand -> Game (Maybe Player)
tableProcessCommandM (DealHands hands) =
  do
    mapM_ (recordEventM . uncurry HandDealt) (Map.toList hands)
    return Nothing
tableProcessCommandM (PlayCard player card) =
  do
    valid <- playValidM player card
    if valid
      then do
        recordEventM (LegalCardPlayed player card)
        turnOverTrick <- turnOverTrickM
        case turnOverTrick of
          Just (trick, trickTaker) ->
            do
              recordEventM (TrickTaken trickTaker trick)
              over <- gameOverM
              case over of
                Just winner ->
                  do
                    recordEventM (GameEnded winner)
                    return (Just winner)
                Nothing ->
                  do
                    recordEventM (PlayerTurnChanged trickTaker)
                    return Nothing
          Nothing ->
            do
              nextPlayer <- playerAfterM player
              recordEventM (PlayerTurnChanged nextPlayer)
              return Nothing
      else do
        recordEventM (IllegalCardAttempted player card)
        return Nothing

-- | Gesamtes Spiel spielen
tableLoopM :: GameCommand -> Game Player
tableLoopM command =
  do
    maybeWinner <- tableProcessCommandM command
    case maybeWinner of
      Nothing ->
        GetCommand tableLoopM
      Just winner ->
        return winner
